In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
import math
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import GridSearchCV, ParameterGrid
from sklearn.ensemble import BaggingRegressor,BaggingClassifier
from sklearn.linear_model import LinearRegression,LogisticRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import roc_curve, precision_recall_curve, auc, make_scorer, recall_score, \
accuracy_score, precision_score, confusion_matrix, mean_squared_error, r2_score
from sklearn.model_selection import StratifiedKFold
from sklearn.tree import DecisionTreeClassifier
import time as time


In [4]:
#import data
X_test = pd.read_csv('X_test.csv')
X_train = pd.read_csv('X_train.csv')
y_train = pd.read_csv('y_train.csv')
y_test = pd.read_csv('y_test.csv')


## Random Forest

### INITAL

In [11]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, accuracy_score

# Create a Random Forest Classifier
rf = RandomForestClassifier(n_estimators=100, random_state=42)

# Train the model using the training sets
rf.fit(X_train, y_train)

# Predict the response for the test dataset
y_pred = rf.predict(X_test)

# Model Accuracy, how often is the classifier correct?
print("Accuracy:", accuracy_score(y_test, y_pred))

# Detailed report of classification done by the model
print(classification_report(y_test, y_pred))


/var/folders/9x/2dgcd7cn3894j3wpmddp46940000gn/T/ipykernel_12858/2575351498.py:8: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  rf.fit(X_train, y_train)


Accuracy: 0.673469387755102
              precision    recall  f1-score   support

           3       0.00      0.00      0.00         8
           4       0.73      0.21      0.32        53
           5       0.69      0.70      0.69       578
           6       0.65      0.78      0.71       887
           7       0.71      0.51      0.59       367
           8       0.96      0.37      0.54        67

    accuracy                           0.67      1960
   macro avg       0.62      0.43      0.48      1960
weighted avg       0.68      0.67      0.66      1960



/opt/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


### Tuning


In [12]:
from sklearn.model_selection import GridSearchCV

# Define the parameter grid
param_grid = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [None, 10, 20, 30, 40],
    'min_samples_split': [2, 10, 20],
    'min_samples_leaf': [1, 5, 10]
}

# Create a GridSearchCV object
grid_search = GridSearchCV(DecisionTreeClassifier(), param_grid, cv=5, scoring='accuracy')

# Fit the GridSearch to our training data
grid_search.fit(X_train, y_train)

print("Best parameters: ", grid_search.best_params_)
print("Best score: ", grid_search.best_score_)


Best parameters:  {'criterion': 'gini', 'max_depth': 40, 'min_samples_leaf': 1, 'min_samples_split': 2}
Best score:  0.550377220734972


In [13]:
from sklearn.metrics import classification_report, confusion_matrix

# Create a new Decision Tree classifier with the best parameters
clf_best = DecisionTreeClassifier(criterion='gini', max_depth=40, min_samples_leaf=1, min_samples_split=2)

# Fit the model on the training data
clf_best.fit(X_train, y_train)

# Predict the labels of the test set
y_pred = clf_best.predict(X_test)

# Print the classification report
print(classification_report(y_test, y_pred))

# Print the confusion matrix
print(confusion_matrix(y_test, y_pred))


              precision    recall  f1-score   support

           3       0.00      0.00      0.00         8
           4       0.23      0.32      0.27        53
           5       0.59      0.61      0.60       578
           6       0.61      0.57      0.59       887
           7       0.48      0.50      0.49       367
           8       0.41      0.43      0.42        67
           9       0.00      0.00      0.00         0

    accuracy                           0.56      1960
   macro avg       0.33      0.35      0.34      1960
weighted avg       0.56      0.56      0.56      1960

[[  0   0   5   2   1   0   0]
 [  0  17  17  14   2   2   1]
 [  0  26 350 166  31   5   0]
 [  1  26 191 510 145  14   0]
 [  1   5  24 133 182  21   1]
 [  0   0   6  12  19  29   1]
 [  0   0   0   0   0   0   0]]


/opt/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/opt/anaconda3/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


- Class 6: Similar to class 5, the precision and recall are around 0.60. This class seems to be the best predicted by the model.
- The accuracy of the model is 0.56 or 56%, which is the proportion of true results (both true positives and true negatives) in the population.
- the model is performing decently for classes 5, 6, and 7 but poorly for the other classes.